src/hotspot/share/classfile/classLoaderExt.cpp

```diff
@@ -284,14 +287,86 @@ InstanceKlass* ClassLoaderExt::load_class(Symbol* name, const char* path, TRAPS)
   }
   stream->set_verify(true);
 
-  ClassLoaderData* loader_data = ClassLoaderData::the_null_class_loader_data();
-  Handle protection_domain;
-  ClassLoadInfo cl_info(protection_domain);
-  InstanceKlass* k = KlassFactory::create_from_stream(stream,
-                                                      name,
-                                                      loader_data,
-                                                      cl_info,
-                                                      CHECK_NULL);
+  InstanceKlass* k = load_class_impl(name, path, CHECK_NULL);
+
+  SystemDictionaryShared::set_shared_class_misc_info(k, (ClassFileStream*)stream);
+
+  return k;
+}
+
+class URLClassLoaderTable : public ResourceHashtable<
```

Comment:
I think we should try to decouple the loading of unregistered classes from ClassLoader/ClassloaderExt. Maybe these new functions can be placed in a new file, shared/cds/unregisteredClasses.cpp, and the API entry point would be `UnregisteredClasses::load_class(name, path)`.

Also, `ClassLoaderExt::find_classpath_entry_from_cache` shouldn't be needed anymore. If the JAR file doesn't exist, URLClassLoader will throw an exception, so we don't need to create a ClassPathEntry just to check for the JAR file's existence.

I wonder if there's any code in classLoader.cpp that was used only for supporting the loading of unregistered classes. If so, it can be removed, too.

Reply:
> I think we should try to decouple the loading of unregistered classes from ClassLoader/ClassloaderExt. Maybe these new functions can be placed in a new file, shared/cds/unregisteredClasses.cpp, and the API entry point would be `UnregisteredClasses::load_class(name, path)`.

I've moved the new code to unregistedClasses.[c|h]pp.
> 
> Also, `ClassLoaderExt::find_classpath_entry_from_cache` shouldn't be needed anymore. If the JAR file doesn't exist, URLClassLoader will throw an exception, so we don't need to create a ClassPathEntry just to check for the JAR file's existence.

As we've discussed off-line, after removing the `ClassLoaderExt::find_classpath_entry_from_cache`, the calculation of the length and crc of the ClassFileSteam will be performed in `jvm_define_class_common()`.
> 
> I wonder if there's any code in classLoader.cpp that was used only for supporting the loading of unregistered classes. If so, it can be removed, too.

I don't see any functions in classLoader.cpp could be removed. However, I think the `ClassLoader::record_result()` could be simplified since it is no longer called with unregistered classes. I'd prefer handle the simplification in a follow-up bug.